src/Pos/Explorer/Web/Server.hs

```diff
@@ -250,6 +298,47 @@ getTxSummary cTxId = do
 -- Helpers
 --------------------------------------------------------------------------------
 
+-- | Find all `MainBlock` by applying the *predicate*, starting from *headerHash*
+filterMainBlocks
+    :: (SscHelpersClass ssc, ExplorerMode m)
+    => HeaderHash
+    -> (MainBlock ssc -> Bool)
+    -> m [MainBlock ssc]
+filterMainBlocks headerHash predicate = rights <$> generalBlockSearch
+  where
+    generalBlockSearch    = filterAllBlocks headerHash specializedPred (pure [])
+    specializedPred block = either (const False) predicate block
+
+-- | Find all blocks matching the sent predicate. This is a generic function
+-- that can be called with either `MainBlock` or `GenesisBlock` in mind.
+filterAllBlocks
+    :: (SscHelpersClass ssc, ExplorerMode m)
+    => HeaderHash
+    -> (Block ssc -> Bool)
+    -> m [Block ssc]
+    -> m [Block ssc]
+filterAllBlocks headerHash predicate acc
```

Comment:
Note that manuall recursions tend to cause memory leaks if they are not strict in the accumulator. I will review it in more details tomorrow but usually similar loops are implemented with fold (which is much safer to reason about memory wise then custom recursion)

Will take a closer look tomorrow